lib/yaml/constructor.py

```diff
@@ -10,16 +10,29 @@
 
 from .error import *
 from .nodes import *
+from .registry import RegistryMeta, ConstructorRegistry
 
+import threading
 import collections.abc, datetime, base64, binascii, re, sys, types
 
 class ConstructorError(MarkedYAMLError):
     pass
 
-class BaseConstructor:
 
-    yaml_constructors = {}
-    yaml_multi_constructors = {}
+class ConstructorSetup(threading.local):
+    def __init__(self):
+        self.initialized = False
+
+    def ensure_initialized(self):
+        if not self.initialized:
+            self.initialized = True
```

Comment:
Are these two lines guaranteed to be atomic? I kind of doubt it - you might need to use a mutex too.

Reply:
Because `ConstructorSetup` and all of the other `*Setup` and `*Registry` classes inherit from `threading.local`, their instance variables are thread-local and `__init__` is called the first time any one of the instance variables is accessed in a new thread. So atomicity isn't really needed here.